core/src/apps/ethereum/definitions.py

```diff
@@ -0,0 +1,142 @@
+        payload = r.read_memoryview(payload_length)
+
+        # at the end compute Merkle tree root hash using
+        # provided leaf data (payload with prefix) and proof
+        hash = sha256(b"\x00" + memoryview(definition)[:r.offset]).digest()
+        proof_length = r.get()
+        for _ in range(proof_length):
+            proof_entry = r.read_memoryview(32)
+            hash_a = min(hash, proof_entry)
+            hash_b = max(hash, proof_entry)
+            hash = sha256(b"\x01" + hash_a + hash_b).digest()
+
+        signed_tree_root = r.read_memoryview(64)
+
+    except EOFError:
+        raise DataError("Invalid Ethereum definition")
+
+    # verify signature
+    if not ed25519.verify(consts.DEFINITIONS_PUBLIC_KEY, signed_tree_root, hash):
+        error_msg = DataError("Invalid definition signature")
+        if __debug__:
+            # check against dev key
+            if not ed25519.verify(
+                consts.DEFINITIONS_DEV_PUBLIC_KEY,
+                signed_tree_root,
+                hash,
+            ):
+                raise error_msg
+        else:
+            raise error_msg
+
+    # decode it if it's OK
+    try:
+        return protobuf.decode(payload, expected_type, True)
+    except ValueError:
+        raise DataError("Invalid Ethereum definition")
+
+
+class Definitions:
+    """Class that holds Ethereum definitions - network and tokens.
+    Prefers built-in definitions over encoded ones.
+    """
+
+    def __init__(
+        self,
+        network: EthereumNetworkInfo = UNKNOWN_NETWORK,
+        tokens: dict[bytes, EthereumTokenInfo] | None = None,
+    ) -> None:
+        self.network = network or UNKNOWN_NETWORK
+        self._tokens = tokens or {}
+
+    @classmethod
+    def from_encoded(
+        cls,
+        encoded_network: bytes | None,
+        encoded_token: bytes | None,
+        chain_id: int | None = None,
+        slip44: int | None = None,
+    ) -> Self:
+        network = UNKNOWN_NETWORK
+        tokens: dict[bytes, EthereumTokenInfo] = {}
+
+        # if we have a built-in definition, use it
+        if chain_id is not None:
+            network = networks.by_chain_id(chain_id)
+        elif slip44 is not None:
+            network = networks.by_slip44(slip44)
+
+        if network is UNKNOWN_NETWORK and encoded_network is not None:
+            network = decode_definition(encoded_network, EthereumNetworkInfo)
+
+        if network is UNKNOWN_NETWORK:
+            # ignore tokens if we don't have a network
+            return cls()
+
+        if chain_id is not None and network.chain_id != chain_id:
+            raise DataError("Network definition mismatch")
+        if slip44 is not None and network.slip44 != slip44:
+            raise DataError("Network definition mismatch")
+
+        # get token definition
+        if encoded_token is not None:
+            token = decode_definition(encoded_token, EthereumTokenInfo)
+            if token.chain_id == network.chain_id:
```

Comment:
there is a weird but plausible usecase with contract decoding / signed data: a bridging contract might talk about tokens on a different chain. in such case, the chain id of the other token would be part of the decoded data (so up to the contract decoder to find and verify) but it would not match the current network.

it's unlikely that we ever add support for anything like this, but ignoring such token definition here instead of raising an error for it will help us if we do